test-utils/runtime/src/substrate_test_pallet.rs

```diff
@@ -0,0 +1,292 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{AccountId, AuthorityId, BlockNumber, Runtime, Signature, Transfer};
+use codec::KeyedVec;
+use frame_support::storage;
+use sp_core::storage::well_known_keys;
+use sp_io::hashing::blake2_256;
+use sp_std::prelude::*;
+
+const NONCE_OF: &[u8] = b"nonce:";
+const BALANCE_OF: &[u8] = b"balance:";
+
+pub use self::pallet::*;
+
+const LOG_TARGET: &str = "substrate_test_pallet";
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {}
+
+	// The current block number being processed. Set by `on_initialize`.
+	#[pallet::storage]
+	pub type Number<T: Config> = StorageValue<_, T::BlockNumber, OptionQuery>;
+
+	#[pallet::storage]
+	pub type NewAuthorities<T> = StorageValue<_, Vec<AuthorityId>, OptionQuery>;
+
+	#[pallet::storage]
+	pub type Authorities<T> = StorageValue<_, Vec<AuthorityId>, ValueQuery>;
+
+	#[pallet::genesis_config]
+	#[cfg_attr(feature = "std", derive(Default))]
+	pub struct GenesisConfig {
+		pub authorities: Vec<AuthorityId>,
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {
+			<Authorities<T>>::put(self.authorities.clone());
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			Number::<T>::put(n);
+			Weight::zero()
+		}
+
+		fn on_finalize(_n: T::BlockNumber) {}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(100)]
+		pub fn authorities_change(
+			origin: OriginFor<T>,
+			new_authorities: Vec<AuthorityId>,
+		) -> DispatchResult {
+			frame_system::ensure_signed(origin)?;
+			<NewAuthorities<Runtime>>::put(new_authorities.to_vec());
+			Ok(())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(100)]
+		pub fn transfer(
+			origin: OriginFor<T>,
+			transfer: Transfer,
+			_signature: Signature,
+			_exhaust_resources_when_not_first: bool,
```

Comment:
This should be just a call `exhaust_resources` with a weight of `T::BlockWeights::get().max_block`.

Or just add the `root-testing` pallet and use `fill_block`: https://github.com/paritytech/substrate/blob/ee38a52d9ea643e4eec1e954b0b603aea4461416/frame/root-testing/src/lib.rs#L44-L53

Reply:
Absolutely doable. Already reworked, example of use: 
https://github.com/michalkucharczyk/substrate/blob/9aadd2ad501e9f09b1e96586082e4b4090fe8f60/client/basic-authorship/src/basic_authorship.rs#L968-L974

However using `root-testing` is not feasible without "heavy" interference into logic of some tests which are using nonce ordering to assume transaction processing order - this is done by using nonce and/or different signers.

So I chose to add [new call](https://github.com/michalkucharczyk/substrate/blob/9aadd2ad501e9f09b1e96586082e4b4090fe8f60/test-utils/runtime/src/substrate_test_pallet.rs#L183-L189) and use `max_block` as weight and do not go with `sudo` and `root-testing` pallets.